src/libmongoc/tests/bsonutil/bson-match.c

```diff
@@ -0,0 +1,592 @@
+                    bson_type_to_string (bson_val_type (actual)));
+         goto done;
+      }
+
+      actual_bson = bson_val_to_array (actual);
+      if (bson_count_keys (expected_bson) != bson_count_keys (actual_bson)) {
+         MATCH_ERR ("expected array of size %" PRIu32
+                    ", but got array of size: %" PRIu32,
+                    bson_count_keys (expected_bson),
+                    bson_count_keys (actual_bson));
+         goto done;
+      }
+
+      BSON_FOREACH (expected_bson, expected_iter)
+      {
+         bson_val_t *expected_val = bson_val_from_iter (&expected_iter);
+         bson_val_t *actual_val = NULL;
+         bson_iter_t actual_iter;
+         const char *key;
+
+         key = bson_iter_key (&expected_iter);
+         if (!bson_iter_init_find (&actual_iter, actual_bson, key)) {
+            MATCH_ERR ("expected array index: %s, but did not exist", key);
+            bson_val_destroy (expected_val);
+            bson_val_destroy (actual_val);
+            goto done;
+         }
+
+         actual_val = bson_val_from_iter (&actual_iter);
+
+         path_child = bson_strdup_printf ("%s.%s", path, key);
+         if (!bson_matcher_match (
+                matcher, expected_val, actual_val, path_child, error)) {
+            bson_val_destroy (expected_val);
+            bson_val_destroy (actual_val);
+            bson_free (path_child);
+            goto done;
+         }
+         bson_val_destroy (expected_val);
+         bson_val_destroy (actual_val);
+         bson_free (path_child);
+      }
+      ret = true;
+      goto done;
+   }
+
+   if (!bson_val_eq (expected, actual, BSON_VAL_FLEXIBLE_NUMERICS)) {
+      MATCH_ERR ("value %s != %s",
+                 bson_val_to_json (expected),
+                 bson_val_to_json (actual));
+      goto done;
+   }
+
+   ret = true;
+done:
+   if (!ret && is_root) {
+      /* Append the error with more context at the root match. */
+      bson_error_t tmp_error;
+
+      memcpy (&tmp_error, error, sizeof (bson_error_t));
+      test_set_error (error,
+                      "BSON match failed: %s\nExpected: %s\nActual: %s",
+                      tmp_error.message,
+                      bson_val_to_json (expected),
+                      bson_val_to_json (actual));
+   }
+   return ret;
+}
+
+bool
+bson_match (bson_val_t *expected, bson_val_t *actual, bson_error_t *error)
+{
+   bson_matcher_t *matcher = bson_matcher_new ();
+   bool matched = bson_matcher_match (matcher, expected, actual, "", error);
+   bson_matcher_destroy (matcher);
+   return matched;
+}
+
+typedef struct {
+   const char *desc;
+   const char *expected;
+   const char *actual;
+   bool expect_match;
+} testcase_t;
+
+static void
+test_match (void)
+{
+   testcase_t tests[] = {
+      {"int32 ==", "{'a': 1}", "{'a': 1}", true},
+      {"int32 !=", "{'a': 1}", "{'a': 0}", false},
+      {"$$exists", "{'a': {'$$exists': true}}", "{'a': 0}", true},
```

Comment:
For completeness' sake, add $$exists false tests

Reply:
Done.